Customizable animation tracks

wschnepp opened the pull request:
- Refactoring so that the plugin can offer an API to specify custom animation track exporters
- API to offer custom animation track exporters

wschnepp committed:
fix the plugin skipping samples

(cherry picked from commit b535539909b567b2f70df921140edcece0710110)

wschnepp committed:
fix the plugin skipping samples

wschnepp committed:
meta files

wschnepp committed:
gitignore

wschnepp committed:
make RemoveUnneededKeyframes internal & static

wschnepp committed:
even better implementation that keeps the memory optimization

(cherry picked from commit 469086d207ef74ae19bbb6897d5b36489fb3e77b)

wschnepp committed:
even better implementation that keeps the memory optimization

wschnepp committed:
update comment

wschnepp committed:
improve comment further

wschnepp committed:
Merge branch 'fix-keyframe-filtering' into fix-keyframe-filtering-on-khronos-dev

# Conflicts:
#	Runtime/Scripts/Timeline/GLTFRecorder.cs

wschnepp committed:
remove random meta files again

Julian24816 committed:
Merge pull request #1 from BII-GmbH/fix-keyframe-filtering-on-khronos-dev

Fix keyframe filtering on khronos dev

wschnepp committed:
Huh why did this suddenly change back?

Julian24816 committed:
Merge pull request #2 from BII-GmbH/oops-fix

Huh why did this suddenly change back?

Commit:
Extract AnimationData to own file

Commit:
Extract ExportPlan & Track

Commit:
begin cleaning up the code

Commit:
moar abstraction hahahaha

Commit:
more refactoring

Commit:
renaming

Commit:
things are broken, but the concept seems fairly sound to me

Commit:
this actually seems to work

Commit:
hmm

Commit:
cleanup

Commit:
small self-review

wschnepp committed:
review feedback + potential bugfix

wschnepp committed:
Merge pull request #3 from BII-GmbH/explicit-visibility-track

Explicit visibility track

robertdorn83 committed:
updated KHR_animatio_pointer to get deserialized

robertdorn83 committed:
added support for KHR_animationpointer import: Nodes, Lights, SkinnendMeshes

robertdorn83 committed:
changed BuildAnimationSamplers to load accessor by accessor data type, not by animation sample type

robertdorn83 committed:
updated KHR_LightsPunctualExtension to support AnimationPointers

robertdorn83 committed:
added material support for animationpointer import (WIP)

robertdorn83 committed:
added support for TextureTransform in animation pointer + some refactoring of material animation pointer (WIP)

robertdorn83 committed:
added support for combined animation samples. eg. for emissive factor and strength. Fefactoring animation pointer structure

robertdorn83 committed:
added camera import for nodes and animationpointer

robertdorn83 committed:
added missing perspective clip plane import for camera animationpointer

robertdorn83 committed:
added Extension Class for GLTFAccessorAttributeType to get component count

robertdorn83 committed:
added conversion to gamme color space for KHR_light_punctual animation pointers

robertdorn83 committed:
fixed innerspotangle on lights when importing and exporting them as an animationpointer

robertdorn83 committed:
refactoring of material mapper, more comments, support for animation property conversion to other property type

robertdorn83 committed:
fixed rebased issues

wschnepp committed:
small doc fixes

wschnepp committed:
Merge branch 'khronos/main' into dev

# Conflicts:
#	Runtime/Scripts/Timeline/GLTFRecorder.cs
#	package.json

wschnepp committed:
rename property back

wschnepp committed:
very experiments

wschnepp committed:
revert experimental changes

wschnepp committed:
revert another experiment

wschnepp committed:
restore random event thing

wschnepp committed:
make clean by creating public facing facade for animation sampler

robertdorn83 committed:
added await YieldOnTimeoutAndThrowOnLowMemory to Import animation

robertdorn83 committed:
changed BuildMaterialAnimationPointerData signature to out animationdata

robertdorn83 committed:
renamed MaterialPointerPropertyMap Type from Generic to Float

robertdorn83 committed:
PointerPath refactoring: renaming and removed PathResolver class

robertdorn83 committed:
refactored AnimationPointerData and AnimPointer Extension interface

robertdorn83 committed:
Added dispersion feature to ShaderGraph (WIP)

Eregerog commented on Runtime/Scripts/Timeline/Internal/AnimationSampler.cs:
```diff
@@ -0,0 +1,105 @@
+                otherSamplers.Add(new BlendWeightSampler());
+            }
+            if (recordAnimationPointer) {
+                // TODO add other animation pointer export plans
```
@wschnepp is this a todo for this PR?

Eregerog replied:
Actually this todo was there biifore, wasn't it...

Eregerog commented on Runtime/Scripts/Timeline/GLTFRecorder.cs:
```diff
@@ -461,13 +296,103 @@ private void CollectAndProcessAnimation(GLTFSceneExporter gltfSceneExporter, GLT
+				else if (scaleTime < visTime) {
+					// the next scale change occurs sooner than the next visibility change
+					// However, if the model is currently invisible, we simply dont care
+                    if (lastVisible) {
+	                    record(scaleTime, scale);
+                    }
+                    scaleIndex++;
```
Identation bii weird

robertdorn83 committed:
reworked IOR mathematics

robertdorn83 committed:
added dispersion extension scripts, added support into material import, export, anim. pointer export

robertdorn83 committed:
added dispersion to MaterialPropertyMapper

robertdorn83 committed:
set roughness default value to 1 (gltf default) in pbrGraph

hybridherbst committed:
fix: avoid overwriting texture files with identical names

robertdorn83 committed:
fixed removing empty root objects when they are animated

robertdorn83 committed:
using GetUniqueName to ensure no image file overwriting with existing paths

robertdorn83 committed:
fixed not correct working Refraction Rend.Feature - cameraColorTarget as source was sometimes just a UnityDefault texture

hybridherbst committed:
simplify GLTFSceneImporter constructors and explicitly define default imported material

hybridherbst committed:
Merge branch 'fix/avoid-overwriting-texture-files' into dev

hybridherbst committed:
revert to default glTF value after checking other implementations

robertdorn83 committed:
added more null checks to GltfRootExtensions

robertdorn83 committed:
code formate changes on MaterialPropertyRemapper > multiples lines for future changes

robertdorn83 committed:
added ClearCoat to Default MaterialPropertyRemapper

robertdorn83 committed:
added roughness TexTransform to Default MaterialPropertiesRemapper

hybridherbst committed:
fix nullreferenceexception when read value from array is not a number

technically invalid but can be encountered on extreme gltf-transform simplification results

hybridherbst committed:
Merge remote-tracking branch 'origin/feature/animation-pointer-import' into dev

hybridherbst committed:
Merge remote-tracking branch 'origin/feature/dispersion' into dev

hybridherbst committed:
Merge remote-tracking branch 'origin/fix/Avoid-Strip-Roots-When-Animated' into dev

hybridherbst committed:
fix null reference when checking for animated root nodes in root stripping

wschnepp committed:
Merge branch 'master' into customizable-animation-tracks

wschnepp committed:
fix plugin exporting (seemingly unused) _Color instead of _BaseColor when using HDRP
Fix color track having the wrong name

hybridherbst committed:
fix: regression where texture transforms are incorrectly applied twice

hybridherbst committed:
some more null checks in JsonExtensions

hybridherbst committed:
fix compilation errors on legacy 2020.x

hybridherbst committed:
changelog/version

hybridherbst committed:
readme and changelog, bump version, demote error on property sampling to warning

TobiasCutik committed:
add individual functions to end recording + create exporter

wschnepp committed:
Merge branch 'khronos-main' into customizable-animation-tracks

wschnepp committed:
check hdrp color property before _Color, as HDRP materials have both, but only one is set

wschnepp committed:
review feedback

Eregerog commented on Runtime/Scripts/Timeline/Internal/Samplers/BlendWeightSampler.cs:
```diff
@@ -6,7 +6,7 @@ internal sealed class BlendWeightSampler : AnimationSampler<SkinnedMeshRenderer,
         public override string PropertyName => "weights";
         
-        protected override SkinnedMeshRenderer getTarget(Transform transform) => transform.GetComponent<SkinnedMeshRenderer>();
+        protected override SkinnedMeshRenderer getTarget(Transform transform) => transform.TryGetComponent<SkinnedMeshRenderer>(out var smr) ? smr : null;
```
nullable annotation

Eregerog commented on Runtime/Scripts/Timeline/Internal/Samplers/BaseColorSampler.cs:
```diff
@@ -7,9 +6,9 @@ internal sealed class BaseColorSampler : AnimationSampler<Material, Color?>
+        public override string PropertyName => "baseColorFactor";
         
-        protected override Material getTarget(Transform transform) => transform.GetComponent<MeshRenderer>()?.sharedMaterial;
+        protected override Material getTarget(Transform transform) => transform.TryGetComponent<MeshRenderer>(out var mr) ? mr.sharedMaterial : null;
```
nullable animation

Eregerog commented on Runtime/Scripts/Plugins/MaterialExtensionsExport.cs:
```diff
@@ -73,7 +74,8 @@ public override void AfterMaterialExport(GLTFSceneExporter exporter, GLTFRoot gl
 			if (!material) return;
 
-			var usesTransmission = material.IsKeywordEnabled("_VOLUME_TRANSMISSION_ON");
+			var usesTransmission = material.IsKeywordEnabled("_VOLUME_TRANSMISSION_ON") || material.IsKeywordEnabled("_VOLUME_TRANSMISSION_ANDDISPERSION");
```
👀

Eregerog commented on Runtime/Scripts/Plugins/AnimationPointerImport.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace UnityGLTF.Plugins
+{
+    public class AnimationPointerImport: GLTFImportPlugin
```
Could maybe bii useful for toughbiias?

Eregerog commented on Runtime/Scripts/Plugins/AnimationPointer/MaterialPropertiesRemapper.cs:
```diff
@@ -0,0 +1,532 @@
+        {
+            var baseColor = new MaterialPointerPropertyMap(MaterialPointerPropertyMap.PropertyTypeOption.SRGBColor)
+            {
+                PropertyNames = new[] { "_Color", "_BaseColor", "_BaseColorFactor", "baseColorFactor" },
```
Maybe these need to be flipped?

wschnepp replied:
not sure what this class even is...but yeah thinking about the other changes probably

Eregerog commented on Runtime/Scripts/Plugins/AnimationPointer/AnimationPointerUtilities.cs:
```diff
@@ -0,0 +1,177 @@
+
+namespace UnityGLTF.Plugins
+{
+    internal static class AnimationPointerHelpers
```
:eyes

Eregerog reviewed (changes requested):
😭

TobiasCutik committed:
skip dataLoader verification when creating a GLTFSceneImporter from stream

TobiasCutik committed:
Merge remote-tracking branch 'origin/customizable-animation-tracks' into cleanup-recorded-glb-export

TobiasCutik committed:
Update GLTFSceneImporter.cs

remove option assignment in public GLTFSceneImporter constructor (as it is already done in the private one)

TobiasCutik committed:
add optional bool verifyDataLoader to GLTFSceneImporter()

TobiasCutik committed:
rename some functions that (partially) end recording

TobiasCutik committed:
simplify code + make renamings for clarity

TobiasCutik committed:
rename TryEndRecordingDirect() to EndRecording()

TobiasCutik committed:
Merge pull request #5 from BII-GmbH/cleanup-recorded-glb-export

Cleanup recorded glb export

TobiasCutik committed:
export only unique node names ensuring that all imported nodes get animated

TobiasCutik committed:
append node count to exported node name only if name is not unique

TobiasCutik committed:
Merge pull request #6 from BII-GmbH/make-only-unique-node-names

make only unique node names

wschnepp committed:
review change - no idea how to even test this though
